### PR TITLE
Fix issue where fixtures would lose the decorated functionality

### LINKED ISSUE #3774: 3.7.1 unittest.mock decorators won't work for pytest fixture
Hi, some of my unittests failed after I upgrade to pytest 3.7.1. I checked them and found it has something to do with using mock decorator with pytest fixtures.

Example:
```
import unittest.mock as mock

import pytest

config = {
    'mykey': '1'
}


@pytest.fixture(scope='function')
@mock.patch.dict(config, {'mykey': 0})
def my_fixture():
    print("value of mykey now: {}".format(config['mykey']))


def test_foobar(my_fixture):
    print('bla bla bla')
```

With pytest 3.6.3, it works fine and output is like:

> ============================= test session starts ==============================
platform darwin -- Python 3.5.0, pytest-3.6.3, py-1.5.4, pluggy-0.6.0
rootdir: /Users/zhangyan/foobar/tests, inifile:
plugins: falcon-0.4.0collected 1 item
test_fixture.py value of mykey now: 0
.bla bla bla

Can see that the value of mykey has been mocked.

But with pytest 3.7.1, the output is:

> ============================= test session starts ==============================
platform darwin -- Python 3.5.0, pytest-3.7.1, py-1.5.4, pluggy-0.7.1
rootdir: /Users/zhangyan/foobar/tests, inifile:
plugins: falcon-0.4.0collected 1 item
test_fixture.py value of mykey now: 1
.bla bla bla

Can see from the output that, the mock didn't make any effect.
Please correct me if my usage is incorrect, thanks!

### PR DESCRIPTION
**Pull request by nicoddemus**
Fix #3774

### CODE REVIEW

**Comment by nicoddemus** [comment]
I don't know how we can make fixtures definitions un-callable, because we will hit the exact same problem we have today with issuing the warning: we still need to wrap the function to raise an error instead of a warning, which will bring the same class of problems. 😕

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus by returning a non-callable object and by raising hard errors if a fixture definition is wrapped with anything else

**Comment by nicoddemus** [comment]
> @nicoddemus by returning a non-callable object and by raising hard errors if a fixture definition is wrapped with anything else

Not sure I follow, won't that bring the exact same problems that we are facing now? I mean, if `@pytest.fixture` returns a non-callable object, the following will blow up:

```python
@classmethod
@pytest.fixture
def setup(cls): ... 
```

Because `@classmethod` will be applied to the object, not to `setup`.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus as far as i can tell its a unreasonable amount of work to "correctly combine the wrappers before and after a fuxture definition markers, with a gernally tremendous chance to be incorrect
so the goal for the breakage is to make any decoration applied to the marker incorrect- the class-method on a fixture definition **should** blow up

**Comment by nicoddemus** [comment]
Hi @RonnyPfannschmidt sorry for the delay on this.

> @nicoddemus as far as i can tell its a unreasonable amount of work to "correctly combine the wrappers before and after a fuxture definition markers, with a gernally tremendous chance to be incorrect

I agree it seems hard to implement.

> so the goal for the breakage is to make any decoration applied to the marker incorrect- the class-method on a fixture definition should blow up

But how would that work for class-scoped fixtures? They by definition need to be `@classmethods`, no? At least that's how it always worked so far.

```python

import pytest

class Test:

    #@classmethod
    @pytest.fixture(scope='class', autouse=True)
    def setup(self):
        self.x = []

    def test1(self):
        assert self.x == []
        self.x.append(1)

    def test2(self):
        assert self.x == [1]
        self.x.append(2)

    def test3(self):
        assert self.x == [1, 2]
```

Without `@classmethod` this never worked as far as I can see, even in previous versions.

**Comment by nicoddemus** [comment]
Unless you mean that we should apply the `@classmethod` decorator ourselves when applying `@pytest.fixture(scope="class")`?

**Comment by RonnyPfannschmidt** [comment]
applying class-method should be done in the correct order - aka 

```
mport pytest

class Test:

   
    @pytest.fixture(scope='class', autouse=True)
    @classmethod
    def setup(cls):
        cls.x = []

    def test1(self):
        assert self.x == []
        self.x.append(1)

    def test2(self):
        assert self.x == [1]
        self.x.append(2)

    def test3(self):
        assert self.x == [1, 2]
```

we should take a look at warning if `lassmethod` is not applied
we shouldn't apply ourselfes

**Comment by nicoddemus** [comment]
> applying class-method should be done in the correct order - aka

Oh I see. We should then advertise this more as the correct way to do the decoration. 

Also I wonder if we might conflict with some other library which also has special requirements? But then again, decorators around fixtures are somewhat rare, and few libraries will return a non-callable object, so I might be overthinking here. 😛 

> we should take a look at warning if lassmethod is not applied
we shouldn't apply ourselfes

Sounds fair, but do you see a possible problem with us applying it ourselves? I'm not suggesting we do, I think the warning or even an error is a good solution, just curious if you are already foreseeing problems with pytest applying the `@classmethod` decorator automatically.

**Comment by RonnyPfannschmidt** [comment]
@nicoddemus its already tricky as is - it would be even worse with multi scoped-fixtures should they ever land - in addition class scope is possibly declared outside - so from my pov there is a massive amount of special cases for next to no gain - warning people to make this right is a lot more easy than correctly setting up things

**Comment by nicoddemus** [comment]
> it would be even worse with multi scoped-fixtures should they ever land

Oh yeah!

> in addition class scope is possibly declared outside - so from my pov there is a massive amount of special cases for next to no gain - warning people to make this right is a lot more easy than correctly setting up things

Sure sounds reasonable. Thanks for the patience to discuss all the points. 👍 

Should we merge this then? It seems like it is an improvement regardless.

File: testing/test_compat.py [hunk 1]
```diff
@@ -26,6 +29,8 @@ def __repr__(self):
             return "<Evil left={left}>".format(left=self.left)
 
         def __getattr__(self, attr):
+            if attr == "__pytest_wrapped__":
```
Comment (RonnyPfannschmidt) [inline testing/test_compat.py]: i just noticed this special case - we shouldn't have it - because in practice this means we blow up on evil objects there

**Pull request merged**